### ISSUE TITLE
can't run daemon.js on 8gb home (starting server)

### ISSUE THREAD

**Issue by cornernote** [alainbryden/bitburner-scripts#22]
It's probably not been a requirement here, but I would like to run the script with a brand new save.  I can probably run some sub-scripts until i get enough to upgrade ram, but for OCD reasons I would like it to run from the start, on it's own.

Below is the output...

```
[home ~/]> free
Total:     8.00GB
Used:      0.00GB (0.00%)
Available: 8.00GB

[home ~/]> run daemon.js 
Running script with 1 thread(s), pid 3 and args: [].
daemon.js: FAIL: (0 retries remaining): Run command returned no pid. Destination: /Temp/owned-source-files.txt.js Command: let result = ""; try { result = JSON.stringify(Object.fromEntries(ns.getOwnedSourceFiles().map(sf => [sf.n, sf.lvl]))); } catch { }
        if (ns.read("/Temp/owned-source-files.txt") != result) await ns.write("/Temp/owned-source-files.txt", result, 'w')
Ensure you have sufficient free RAM to run this temporary script.
```

I'm trying to understand what's happening to take the RAM.

I can see the file...

`nano /Temp/owned-source-files.txt.js`

```js
import { formatMoney, formatNumberShort, formatDuration, parseShortNumber, scanAllServers } from "helpers.js"
export async function main(ns) { try { let result = ""; try { result = JSON.stringify(Object.fromEntries(ns.getOwnedSourceFiles().map(sf => [sf.n, sf.lvl]))); } catch { }
        if (ns.read("/Temp/owned-source-files.txt") != result) await ns.write("/Temp/owned-source-files.txt", result, 'w'); } catch(err) { ns.tprint(String(err)); throw(err); } }
```

From this I understand that [ns.getOwnedSourceFiles()](https://github.com/danielyxie/bitburner/blob/dev/markdown/bitburner.ns.getownedsourcefiles.md) has a 5GB RAM cost.

I can see that `daemon.js` has a 5.5GB RAM cost, by doing this:

```
[home ~/]> run daemon.js --tail
Running script with 1 thread(s), pid 5 and args: [].
[home ~/]> free
Total:     8.00GB
Used:      5.50GB (68.75%)
Available: 2.50GB
```

-- Random Question -- Is there a better way to see the RAM cost of a script?

This means `daemon.js` would have to cost 3GB RAM or less in order to execute on the starter 8GB machine.

Has this been tried already?  Do you have any thoughts/advice?

**Comment by cornernote** [comment]
random question answer...

```
[home ~/]> mem daemon.js
This script requires 5.50GB of RAM to run for 1 thread(s)
```

**Comment by cornernote** [comment]
also,

```
[home ~/]> mem /Temp/owned-source-files.txt.js
This script requires 6.60GB of RAM to run for 1 thread(s)
```

**Comment by alainbryden** [comment]
I'll test it later, but I designed tryGetOwnedSourceFiles so that if it
fails (e.g. due to insufficient RAM) it'll raise an error message, but
default to assuming you have no owned source files, and everything should
continue working on that basis.

Is it actually crashing, or just raising lots of warnings/errors?

**Comment by alainbryden** [comment]
In fact, I was missing a try-catch to recover from the error after 5 failed attempts, so now it should still give warnings / an error indicating that it couldn't run, but it should suppress the actual exception and allow all programs to continue to run normally under the assumption that there are no owned source files.

It may keep happening periodically until RAM is expanded. If it's too much, I can handle the excessive noise separately in the daemon by caching the value once it's been returned once (if it's not already cached - I think it might be)

**Comment by cornernote** [comment]
tried again now... it gave some RAM toast messages, but daemon is running!  :)

```
[home ~/]> run daemon.js 
Running script with 1 thread(s), pid 2 and args: [].
[home ~/]> ps
(PID - 2) daemon.js
(PID - 12) hacknet-upgrade-manager.js -c --max-payoff-time 1h
```

**Comment by cornernote** [comment]
Ummm seems we can use more RAM than we have...

```
[home ~/]> free
Total:             8.00GB
Used:             11.60GB (145.00%)
Available: -3600000000.00
[home ~/]> top
Script                                  PID       Threads         RAM Usage
daemon.js                               2         1               5.50GB
hacknet-upgrade-manager.js              12        1               6.10GB
```

Also, this message is the main noise.  Can it be run only if we have >8GB ?

```
daemon.js: FAIL: (0 retries remaining): Run command returned no pid. Destination: /Temp/sell-hashes-for-money.js Command: 0; if(ns.hacknet.spendHashes("Sell for Money")) ns.toast('Sold 4 hashes for $1M', 'success')
Ensure you have sufficient free RAM to run this temporary script.
```

**Comment by alainbryden** [comment]
I can tidy that up, it should only run if hacknet servers are available

**Comment by alainbryden** [comment]
Just wanted you to know that my latest changes make an effort to postpone a few scripts from starting up until there's sufficient RAM. 8GB is probably still too little.

I also fixed the hacknet issue mentioned in the comments.

**Comment by cornernote** [comment]
I'd like to thank you for sharing your code.  I have learned so much from it!

After taking inspiration from your code, I wrote something that runs in 8GB... it can root servers, hack for profit, and manage hacknet!  It uses your concept of ns.write(random.js) to run pretty much every paid function in the background.

Have a play if you'd like to...

[bitburner online](https://danielyxie.github.io/bitburner/)

```
wget https://raw.githubusercontent.com/cornernote/bitburner-scripts/main/src/install.js install.js
run install.js
work
```

The key is this class -

https://github.com/cornernote/bitburner-scripts/blob/main/src/lib/Runner.js

It lets you write code like this...

```
let server = await runner.nsProxy['getServer']('n00dles') 
```

See a really simple example here:

https://github.com/cornernote/bitburner-scripts/blob/main/src/samples/get-player-data.js
